torchao/csrc/metal/int4_quantized_kernels.metal

```diff
@@ -0,0 +1,128 @@
+  n = n * 4;
+  uint k = (tid_in_simdgroup % threads_per_channel) * ks_per_thread;
+  constexpr int k_jump = threads_per_channel * ks_per_thread;
+
+  using vecT = typename Vec4Type<T>::type;
+  constant vecT *A_ptr = reinterpret_cast<constant vecT *>(A);
+  constant uchar *B_ptr = B + ((n * K) / k_pack_factor);
+
+  thread float4 rc = float4(0.0);
+  float4 act_div_scales = {1.f, 1 / 16.f, 1 / 256.f, 1 / 4096.f};
+
+  uint k_block_index = k / groupSize;
+  uint scales_n_offset = (k_block_index * N + n) * 2;
+  uint zeros_n_offset = scales_n_offset + 1;
+  uint scales_jump =
+      N * 2 *
+      (k_jump /
+       groupSize); /* the last term accounts for identifying the group this
+                      thread will have to process in each iteration. This mean
+                      each iteration it must jump to a different group. Thus
+                      k_jump must be > grupSize */
+  for (; k < K; k += k_jump) {
+    const T scale0 = scalesAndZeros[scales_n_offset];
+    // Adding zero point results in 10% perf penalty.
+    const T zero0 = scalesAndZeros[zeros_n_offset] - scale0 * T(8);
+
+    const T scale1 = scalesAndZeros[scales_n_offset + 2];
+    const T zero1 = scalesAndZeros[zeros_n_offset + 2] - scale1 * T(8);
+
+    const T scale2 = scalesAndZeros[scales_n_offset + 4];
+    const T zero2 = scalesAndZeros[zeros_n_offset + 4] - scale2 * T(8);
+
+    const T scale3 = scalesAndZeros[scales_n_offset + 6];
+    const T zero3 = scalesAndZeros[zeros_n_offset + 6] - scale3 * T(8);
+
+    scales_n_offset += scales_jump;
+    zeros_n_offset += scales_jump;
+
+    const float4 zeros = float4(zero0, zero1, zero2, zero3);
+
+    float4 a_val = float4(A_ptr[k / 4]);   // * k_scales;
+    float4 a_vec = a_val * act_div_scales; // * k_scales;
+    float a_val_sum = a_val[0] + a_val[1] + a_val[2] + a_val[3];
+
+    float4x4 b_mat;
+    ushort b_val0 = (reinterpret_cast<constant ushort *>(
+        B_ptr + (k + 0 * K) / k_pack_factor))[0];
+    ushort b_val1 = (reinterpret_cast<constant ushort *>(
+        B_ptr + (k + 1 * K) / k_pack_factor))[0];
+    ushort b_val2 = (reinterpret_cast<constant ushort *>(
+        B_ptr + (k + 2 * K) / k_pack_factor))[0];
+    ushort b_val3 = (reinterpret_cast<constant ushort *>(
+        B_ptr + (k + 3 * K) / k_pack_factor))[0];
+    b_mat[0] = scale0 * float4(float(b_val0 & 0x000f), float(b_val0 & 0x00f0),
+                               float(b_val0 & 0x0f00), float(b_val0 & 0xf000));
+    b_mat[1] = scale1 * float4(float(b_val1 & 0x000f), float(b_val1 & 0x00f0),
+                               float(b_val1 & 0x0f00), float(b_val1 & 0xf000));
+    b_mat[2] = scale2 * float4(float(b_val2 & 0x000f), float(b_val2 & 0x00f0),
+                               float(b_val2 & 0x0f00), float(b_val2 & 0xf000));
+    b_mat[3] = scale3 * float4(float(b_val3 & 0x000f), float(b_val3 & 0x00f0),
+                               float(b_val3 & 0x0f00), float(b_val3 & 0xf000));
+
+    rc += a_vec * b_mat;
```

Comment:
what if k % 4 != 0?

Reply:
it wont work. Overall we will have specialized shaders that are fast for specific shapes. Integration code will either throw or dispatch to appropriate shader which are fallback when we have them